File: android/src/main/kotlin/fl/channel/FlEvent.kt
package fl.channel

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

class FlEvent(
    name: String, messenger: BinaryMessenger,
) {
    private var eventSink: EventChannel.EventSink? = null
    private var eventChannel: EventChannel = EventChannel(messenger, name)
    private val handler = Handler(Looper.getMainLooper())
    private val streamHandler = object : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            eventSink = events
        }

        override fun onCancel(arguments: Any?) {
            cancel()
        }
    }

    init {
        setStreamHandler()
    }

    fun setStreamHandler() {
        eventChannel.setStreamHandler(streamHandler)
    }

    fun cancel() {
        eventSink?.endOfStream()
        eventSink = null
        eventChannel.setStreamHandler(null)
    }


    fun send(args: Any): Boolean {
        if (eventSink != null) {
            handler.post {
                eventSink?.success(args)
            }
        }
        return eventSink != null
    }

}

File: macos/Classes/FlEvent.swift
import FlutterMacOS
import Foundation

public class FlEvent: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var eventChannel: FlutterEventChannel?
    private var binaryMessenger: FlutterBinaryMessenger

    public init(_ name: String, _ binaryMessenger: FlutterBinaryMessenger) {
        self.binaryMessenger = binaryMessenger
        eventChannel = FlutterEventChannel(name: name, binaryMessenger: binaryMessenger)
        super.init()
        eventChannel!.setStreamHandler(self)
    }

    public func send(_ args: Any?) -> Bool {
        if eventSink != nil {
            DispatchQueue.main.async {
                self.eventSink?(args)
            }
        }
        return eventSink != nil && eventChannel != nil
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        eventChannel?.setStreamHandler(nil)
        return nil
    }
}
